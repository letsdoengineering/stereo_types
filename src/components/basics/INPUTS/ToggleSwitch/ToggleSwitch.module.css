/* The whole toggle */
.container {
    position: relative;
    display: inline-block;
    width: 2rem;
    height: 1rem;
}

/* Hide default HTML checkbox */
.container input {
    opacity: 0;
    width: 0;
    height: 0;
}

/* unchecked style */
.slider {
    position: absolute;
    cursor: pointer;
    border-radius: 1rem;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: var(--outline-gray);
    -webkit-transition: .4s;
    transition: .4s;
}

/* unchecked style */
.slider:before {
    border-radius: 50%;
    position: absolute;
    content: "";
    height: 1rem;
    width: 1rem;
    left: 0;
    bottom: 0;
    background-color: white;
    border: 1px solid var(--outline-gray);
    -webkit-transition: .4s;
    transition: .4s;
}

.container input:focus + .slider {
    -moz-box-shadow:    0 0 0 2px var(--focus);
    -webkit-box-shadow: 0 0 0 2px var(--focus);
    box-shadow:         0 0 0 2px var(--focus);
}

/* checked style */
.container input:checked + .slider {
    background-color: var(--highlight-blue);
}

/* checked style */
.container input:checked + .slider:before {
    border: 1px solid var(--light-blue);
    background-color: var(--light-blue);
    -webkit-transform: translateX(1rem);
    -ms-transform: translateX(1rem);
    transform: translateX(1rem);
}
