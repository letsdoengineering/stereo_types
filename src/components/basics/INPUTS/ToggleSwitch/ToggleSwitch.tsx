import React from 'react'
import * as styles from './ToggleSwitch.module.css'

type ToggleSwitchProps = {
  /** Tell the toggle (checkbox) it's status; controlled by consumer via passing onChange function */
  isChecked: boolean
  /** Called when clicked */
  onChange(): void
  /** Text label for screen reader, visually not shown */
  label: string
}

const ToggleSwitch: React.FC<ToggleSwitchProps> = ({
  isChecked,
  onChange,
  label,
  ...rest
}: ToggleSwitchProps) => {
  return (
    <label className={styles.container}>
      <span className='visually-hidden'>{label}</span>
      <input type='checkbox' checked={isChecked} onChange={onChange} {...rest} />
      <span className={styles.slider} />
    </label>
  )
}

export default ToggleSwitch
