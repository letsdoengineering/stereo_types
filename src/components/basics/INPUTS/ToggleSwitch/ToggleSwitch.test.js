import * as React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import { axe } from 'jest-axe'

import ToggleSwitch from './ToggleSwitch'

describe('ToggleSwitch', () => {
    it('should call onChange when clicked', () => {
        const onChange = jest.fn()
        render(<ToggleSwitch isChecked={false} onChange={onChange} label={'screen reader label'} />)
        const toggle = screen.getByLabelText('screen reader label')
        userEvent.click(toggle)
        expect(onChange).toHaveBeenCalled()
    })

    it('should render correctly', () => {
        const onChange = jest.fn()
        const tree = renderer
            .create(
                <ToggleSwitch isChecked={true} onChange={onChange} label={'screen reader label'} />
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders the page DOM that is accessible', async () => {
        const onChange = jest.fn()
        const { container } = render(
            <ToggleSwitch isChecked={false} onChange={onChange} label={'screen reader label'} />
        )
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})
