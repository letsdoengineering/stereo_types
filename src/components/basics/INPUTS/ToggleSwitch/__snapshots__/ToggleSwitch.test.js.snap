// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch should render correctly 1`] = `
<label
  className="container"
>
  <span
    className="visually-hidden"
  >
    screen reader label
  </span>
  <input
    checked={true}
    onChange={[MockFunction]}
    type="checkbox"
  />
  <span
    className="slider"
  />
</label>
`;
